flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,87 +75,360 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
+        mlir::arith::ConstantOp lb = rewriter.create<mlir::arith::ConstantOp>(
+            liveIn.getLoc(), extentType,
+            rewriter.getIntegerAttr(extentType, 0));
+        // TODO I think this caluclation might not be correct. But this is how
+        // it is done in PFT->OpenMP lowering. So keeping it like this until we
+        // double check.
+        mlir::Value ub =
+            rewriter.create<mlir::arith::SubIOp>(liveIn.getLoc(), extent, lb);
+
+        return rewriter.create<mlir::omp::MapBoundsOp>(
+            liveIn.getLoc(), rewriter.getType<mlir::omp::MapBoundsType>(), lb,
+            ub, extent, mlir::Value{}, false, mlir::Value{});
+      };
+
+      for (auto extent : extents)
+        result.push_back(genBoundsOp(extent));
+
+      return result;
+    }();
+
+    return Fortran::lower::omp::createMapInfoOp(
+        rewriter, liveIn.getLoc(), declareOp.getBase(), /*varPtrPtr=*/{},
+        declareOp.getUniqName().str(), boundsOps, /*members=*/{},
+        static_cast<
+            std::underlying_type_t<llvm::omp::OpenMPOffloadMappingFlags>>(
+            mapFlag),
+        captureKind, liveInType);
+  }
+
+  mlir::omp::TargetOp genTargetOp(mlir::Location loc,
+                                  mlir::ConversionPatternRewriter &rewriter,
+                                  mlir::IRMapping &mapper,
+                                  const llvm::SmallVector<mlir::Value> &liveIns,
+                                  mlir::omp::TargetClauseOps &clauseOps) const {
+    mlir::omp::TargetOp targetOp =
+        rewriter.create<mlir::omp::TargetOp>(loc, clauseOps);
+
+    genBodyOfTargetOp(rewriter, targetOp, liveIns, clauseOps.mapVars, mapper);
+    return targetOp;
+  }
+
+  void genBodyOfTargetOp(mlir::ConversionPatternRewriter &rewriter,
+                         mlir::omp::TargetOp targetOp,
+                         const llvm::SmallVector<mlir::Value> &liveIns,
+                         const llvm::SmallVector<mlir::Value> &liveInMapInfoOps,
+                         mlir::IRMapping &mapper) const {
+    mlir::Region &region = targetOp.getRegion();
+
+    llvm::SmallVector<mlir::Type> liveInTypes;
+    llvm::SmallVector<mlir::Location> liveInLocs;
+
+    for (mlir::Value liveIn : liveIns) {
+      liveInTypes.push_back(liveIn.getType());
+      liveInLocs.push_back(liveIn.getLoc());
     }
-    // ==== TODO (2) End ====
-    //
-    // TODO (1 & 2): Isolating analyses proposed in both TODOs, I think we can
-    // more easily generalize the pass to work for targets other than OpenMP,
-    // e.g. OpenACC, I think can, can reuse the results of the analyses and only
-    // change the code-gen/rewriting.
 
-    mlir::IRMapping mapper;
+    rewriter.createBlock(&region, {}, liveInTypes, liveInLocs);
 
-    // Collect the memref defining ops in the parallel region.
-    for (mlir::Operation *opToClone : declareAndAllocasToClone) {
-      rewriter.clone(*opToClone, mapper);
+    for (auto [arg, mapInfoOp] :
```

Comment:
seems reasonable from what I can see, but I would also keep in mind that during the generation of Map <-> BlockArgs that occurs, we also have a little segment of code that clones bounds into the target region where possible (I think genLiveInDeclare may be trying to do something similar but please do correct me if I am wrong), and if it can't I believe it generates a map for the bounds, but @TIFitis is more familiar with this and can perhaps give further input or answer questions you might have on it!